keystorage/yubihsm2/src/test/java/tech/pegasys/signers/yubihsm/YubiHsmSessionTest.java

```diff
@@ -0,0 +1,115 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.signers.yubihsm;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+import tech.pegasys.signers.yubihsm.backend.YubiHsmBackend;
+import tech.pegasys.signers.yubihsm.backend.YubiHsmBackendTestFactory;
+import tech.pegasys.signers.yubihsm.exceptions.YubiHsmException;
+import tech.pegasys.signers.yubihsm.model.Opaque;
+
+import java.nio.file.Path;
+
+import io.specto.hoverfly.junit5.HoverflyExtension;
+import io.specto.hoverfly.junit5.api.HoverflyConfig;
+import io.specto.hoverfly.junit5.api.HoverflySimulate;
+import org.apache.tuweni.bytes.Bytes;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+
+@Nested
+@HoverflySimulate(
+    enableAutoCapture = true,
+    config =
+        @HoverflyConfig(captureAllHeaders = true, proxyLocalHost = true, statefulCapture = true))
+@ExtendWith(HoverflyExtension.class)
+class YubiHsmSessionTest {
+
+  @Test
+  void authenticateSessionAndGetOpaqueData(@TempDir Path tempDir) {
+    final YubiHsmBackend backend = YubiHsmBackendTestFactory.createYubiHsmBackend(tempDir);
+
+    try (final YubiHsmSession yubiHsmSession =
+        spy(new YubiHsmSession(backend, (short) 1, "password".toCharArray()))) {
+      when(yubiHsmSession.getHostChallenge()).thenReturn(Bytes.fromHexString("aaaabbbbccccdddd"));
+
+      yubiHsmSession.authenticateSession();
+
+      final Bytes expected =
+          Bytes.fromHexString("0x5e8d5667ce78982a07242739ab03dc63c91e830c80a5b6adca777e3f216a405d");
+
+      final Bytes key1 = new Opaque((short) 15).getOpaque(yubiHsmSession, Opaque.OutputFormat.HEX);
+      assertThat(key1).isEqualTo(expected);
+
+      final Bytes key2 =
+          new Opaque((short) 16).getOpaque(yubiHsmSession, Opaque.OutputFormat.ASCII);
+      assertThat(key2).isEqualTo(expected);
+    }
+  }
+}
+
+@Nested
+@HoverflySimulate(
+    enableAutoCapture = true,
+    config =
+        @HoverflyConfig(captureAllHeaders = true, proxyLocalHost = true, statefulCapture = true))
+@ExtendWith(HoverflyExtension.class)
+class YubiHsmInvalidAuthenticateSessionTest {
+
+  @Test
+  void invalidCredentialsThrowsException(@TempDir Path tempDir) {
+    final YubiHsmBackend backend = YubiHsmBackendTestFactory.createYubiHsmBackend(tempDir);
+
+    try (final YubiHsmSession session =
+        spy(new YubiHsmSession(backend, (short) 1, "invalidpassword".toCharArray()))) {
+      when(session.getHostChallenge()).thenReturn(Bytes.fromHexString("aaaabbbbccccdddd"));
```

Comment:
it can be dangerous to mock out parts of the unit under test - is there anyway around doing this?

Reply:
we can probably subclass YubiHsmSession and override getHostChallenge to achieve the same result.